docs/contributing/999-rfcs/2023-10-01-workloads.md

```diff
@@ -0,0 +1,252 @@
+`tf-aws`), we should treat this like we treat publishing an AWS Lambda zip archives to an S3 bucket
+during deployment. Platform providers will need to explore a variety of configuration options to
+control this behavior (P2).
+
+We might be able to leverage the [Docker provider for
+Terraform](https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs) with the approach
+[suggested here](https://stackoverflow.com/a/76608435) to push the image during deployment to a
+container registry.
+
+>
+> **Implementation Note:** Design the publishing mechanism with heavy hash-based caching in mind so
+> that both `wing compile` and the Terraform image publishing step will be no-op if the container
+> image hasn't changed.
+>
+
+### Container runtime environment
+
+> Reqtag: `c:environment`
+
+It should be possible to configure the runtime environment of the containers.
+
+The following is a non-exhaustive list of capabilities we will need to support:
+
+* Expose network ports of various types (http, http2, grpc, ??) (P1)
+* Command line arguments to pass to the container (P1)
+* Environment variables (P1)
+* Entrypoint command to override (if you wish to use a pre-built container image and just change the
+  entrypoint) (P1)
+* CPU and memory requirements (in some normalized model such as vCPUs, GBs) (P2)
+* GPUs (P2)
+* Liveness probe
+* Readiness probe (P1)
+* Sidecar containers (P2)
+* Volumes (P2)
+* Autoscaling (P1)
+* etc...
+
+Additional work is required to determine the correct surface area for this API. We recommend to try
+and identify the common denominator across multiple container-based platforms such as Kubernetes
+Deployment + Service, [ECS](https://aws.amazon.com/ecs/), [fly.io](https://fly.io),
+[ControlPlane](https://github.com/controlplane-com/terraform-provider-cpln/blob/main/docs/resources/workload.md),
+[Porter](https://www.porter.run/), etc.
+
+### API endpoints (P1)
+
+> Reqtag: `c:endpoints`
+
+Users must be able to define path-based routing rules so that requests will be routed to the
+workload.
+
+Common underlying providers for this are the various Application Load Balancers which are usually
+configured via Kubernetes [ingress
+rules](https://kubernetes.io/docs/concepts/services-networking/ingress/).
+
+## Cloud implementations
+
+> Reqtag: `c:backends`
+
+Wing platform providers should be able to use one of these container orchestration systems as a
+backend to run containerized workloads:
+
+1. EKS/GKE/AKS (P1)
+2. [ECS](https://aws.amazon.com/ecs/)
+3. Helm (basically Kubernetes `Deployment`, `Service` and `Ingress` resources)
+4. [fly.io](https://fly.io)
+5. [ControlPlane](https://github.com/controlplane-com/terraform-provider-cpln/blob/main/docs/resources/workload.md#nestedblock--options)
+
+The default behavior for `tf-aws` should be to create a dedicated EKS cluster for the app (P1) but
+users should be able to point to an existing EKS cluster through platform configuration options
+(P2).
```

Comment:
Why kubernetes as a default? If you're targeting AWS, it would make sense to use their primary container orchestrator. ECS would be faster, simpler, cheaper, and still sufficiently powerful.

It probably seems like doing EKS/GKE/AKS is like killing 3 birds with 1 stone, but I'm not convinced there will be much shared work between the three.

Reply:
Kubernetes is by far the most common way to run containers on the cloud, both based on data and our conversations with customers.

Indeed, the cluster provisioning is very different across managed k8s platforms, but the actual k8s resources are going to be the same.

Having said that, like you said, it shouldn't be too hard to deploy workloads into ECS and I believe we will support this use case quite quickly.

I think the more common scenario will actually be to reference an existing cluster instead of provisioning one for each wing app, but we want wing standard platforms to work out of the box.

Reply:
> Indeed, the cluster provisioning is very different across managed k8s platforms, but the actual k8s resources are going to be the same.

I don't think this is true. In my experience, doing EKS properly means resources will have aws-specific things. This ranges from simple annotations and sidecars to setting up specific [controllers](https://github.com/kubernetes-sigs/aws-load-balancer-controller/).
It's been a few years and it looks like they've done some work to build in some of this to EKS, so that's good, but the manifests definitely won't be exactly the same.

I know this is just using k8s for the sake of k8s so there's no changing this path, just wanted to keep voicing my concerns. I think k8s is extremely hard to do correctly wrt security and resilience and it will be very easy for us to hand over a terribly configured resources since that's the default k8s experience. Think this is fine if the user wants to explicitly opt-in to a kubernetes experience, but if they just wanted to run a container then defaulting them to this is doing a disservice imo.

Reply:
> doing EKS properly means resources will have aws-specific things

You are right, the manifests won't be exactly the same. There could be some specific additions to the K8S resources synthesized based on the target platform, but the general setup is going to be similar.

> I think k8s is extremely hard to do correctly wrt security and resilience and it will be very easy for us to hand over terribly configured resources since that's the default k8s experience

That's another good reason to dive in head first and not punt these challenges.

> defaulting them to this is doing a disservice imo

I see what you mean. We can discuss what makes sense as the default backend for `cloud.Workload` on AWS down the road (it's totally possible that ECS Fargate will make more sense). We need to prioritize one or the other for our MVP, and there's a good case for starting with Kubernetes, especially given there will be some reuse across providers.

I'll tweak the wording.

Reply:
@chriscbr I added a "Cluster" section with some details. Let me know if this requires additional info.